source/common/upstream/wrsq_scheduler.h

```diff
@@ -114,7 +122,11 @@ class WRSQScheduler : public Scheduler<C>, protected Logger::Loggable<Logger::Id
     maybeRebuildCumulativeWeights();
 
     const double weight_sum = cumulative_weights_.back().cumulative_weight;
-    uint64_t rnum = random_.random() % static_cast<uint32_t>(weight_sum);
+
+    // It's possible that the weight sum can be fractional, so we need a random double.
+    const double r = static_cast<double>(random_.random() % accuracy()) / accuracy();
+    const double rnum = r * weight_sum;
```

Comment:
I think this algorithm breaks down in cases weight_sum > accuracy() / N.  I'm not sure how big weight_sum would need to be for problems to be noticiable.  weight_sum > accuracy() would result some elements being impossible to represent.

weight_sum does involve a sum of weight_val * it.second.size(), so I would expect weight_sum to be relatively large in cases where there is a large number of hosts.

Consider that the original algorithm that does mod weight_sum will produce exactly the same result as the new algorithm in cases where the test random source returns consecutive integers starting at 0.

Reply:
Oh wow, that's a good catch. I'll need to think about the best way to go about this..